tao_compiler/mlir/disc/transforms/fusion_utils.cc

```diff
@@ -2521,5 +2822,1033 @@ bool StitchCPUAnalysis::doCodeGeneration(OpBuilder& b, lmhlo::FusionOp fusion) {
+        if ((dominant_tile.tileSizes.count(lhs[i]) == 0) != non_tiled) {
+          // Both tiled and non-tiled dims are mapped to the same dim of res,
+          // thus cannot determine the tile plan for res;
+          irregular_xroots.insert(op);
+          irregular = true;
+          break;
+        }
+      }
+      if (irregular) {
+        break;
+      }
+      if (non_tiled) {
+        res_non_tiled_dims.insert(eq_dim.second.begin(), eq_dim.second.end());
+        dominant_non_tiled_dim_matched_n += lhs.size();
+      }
+    }
+    if (irregular) {
+      continue;
+    } else if (dominant_non_tiled_dim_matched_n != 1) {
+      // Check constraint: the element-number of non-tiled dims are the same
+      // between sub-roots and regular xroots. If not every non-tiled dim is
+      // matched between sub-root and current op, it is an irregular xroot. Note
+      // we know that sub-root is reduction and has only 1 non-tiled dim.
+      irregular_xroots.insert(op);
+      continue;
+    } else if (*std::max_element(res_non_tiled_dims.begin(),
+                                 res_non_tiled_dims.end()) !=
+               res_non_tiled_dims.size() - 1) {
+      // Check constraint: the tiled dims are all minor dims for all regular
+      // xroot ops. This means the non-tiled dims should all be the major dims.
+      irregular_xroots.insert(op);
+      continue;
+    }
+
+    auto& plan = tile_plan[res];
+    int64_t rank = res.getType().cast<MemRefType>().getRank();
+    for (int64_t i = 0; i < rank; i++) {
+      if (res_non_tiled_dims.count(i) == 0) {
+        plan.tileSizes[i] = ShapedType::kDynamicSize;
+      }
+    }
+    regular_xroots.insert(op);
+  }
+
+  // Check constraint: all the external-only-roots should be regular. This is
+  // because all external-only-roots are skeleton ops. It requires all skeleton
+  // ops have the same non-tiling element numbers to ease the building of loop
+  // loop skeleton of the fusion during code generation.
+  for (auto extern_only_res : fusion_pattern.getExternalOnlyResults()) {
+    Operation* op = fusion_pattern.findLastWriter(extern_only_res);
+    if (irregular_xroots.contains(op)) {
+      LLVM_DEBUG(llvm::dbgs() << "Invalid irregular xroot: " << *op);
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool StitchGpuFusionStrategy::backtraceTileAndCover(
+    ShapeAnalysis& shapeAnalysis, FusionPattern& fusion_pattern, Value value) {
+  auto op_list = fusion_pattern.getOpList();
+  auto subroots = fusion_pattern.getSubRootOps();
+  auto roots = fusion_pattern.getRootOps();
+  DenseMap<Value, TileInfo>& tile_plan = fusion_pattern.getTilePlan();
+  DenseSet<Operation*> op_set(op_list.begin(), op_list.end());
+  DenseSet<Operation*> subroots_set(subroots.begin(), subroots.end());
+  DenseSet<Operation*> roots_set(roots.begin(), roots.end());
+  std::function<bool(Value, bool&)> propagateO2I;
+  propagateO2I = [&](Value value, bool& cover) {
```

Comment:
nit: just
`auto propagateO2I = [&](... `should be enough for short

Reply:
The `propagateO2I` function is recursive, which requires a predefined function prototype.